### ISSUE TITLE
Is there any way to get the same results in different versions?

### ISSUE THREAD

**Issue by Namo-Bang** [dmlc/xgboost#6679]
How do I get the same results in versions 0.90 and 1.0.0? The model trained with 0.90 cannot be used for shap, so I ask.
I really did my best to get the same results in both versions, but I couldn't do it in the end. I need a model that can be used for shap while producing exactly the same results as version 0.90 for the current project. Is there any way?

**Comment by trivialfis** [comment]
Could you please share your model or the data to train the model?

**Comment by trivialfis** [comment]
It should be reproducible aside from possible floating point error.

**Comment by Namo-Bang** [comment]
@trivialfis I cannot share my data due to rule of corp. I can only share params that I set.
```
xgb.XGBClassifier(colsample_bytree=0.3755946915662365, eval_metric='auc', learning_rate=0.01352299840726828, max_depth=7, n_estimators=10000, n_jobs=-1, objective='binary:logistic', reg_alpha=1.0599388680443669, reg_lambda=1.0419613042830034, tree_method='gpu_hist')
```

**Comment by trivialfis** [comment]
@Namo-Bang The gpu hist before 1.0 is not deterministic in training due to floating point arithmetic, see https://xgboost.readthedocs.io/en/latest/parameter.html#additional-parameters-for-gpu-hist-tree-method .  We tackled it in later version.

**Comment by trivialfis** [comment]
The doc was being conservative, so far the accuracy is largely unaffected.

**Comment by trivialfis** [comment]
I can't recall the exact version we added that parameter.  You may have to do a bit lookup in release news.

**Comment by Namo-Bang** [comment]
@trivialfis Even if I set the deterministic_histogram parameter to False, it still learns differently from version 0.90. Are there any additional parameters that need to be set?

**Comment by trivialfis** [comment]
There will be slight difference due to floating point error.  For 0.90, it can be different even if you are just running it twice using the exact same data and libraries.  After adding that parameter, the guarantee is the result should be bit identical if you are using same release.  So there's no point in reproducing the model from 0.90, as it can conflict with itself during 2 runs due to floating point error.

**Comment by trivialfis** [comment]
Another issue is like: https://github.com/dmlc/xgboost/issues/6350

**Comment by Namo-Bang** [comment]
@trivialfis 
Thank you so much for continuing to answer sincerely. One last question I want to ask.

My issue is not about load_model or  floating point error but train model. When I load model trained with version 0.90 in version 1.x.x, I cannot use this model for shap library. So I have to train model again in version 1.x.x. The issue is that I can not get same model, even if all parameters are same all data are same all seeds are same. 

In my thought, It seems that there are critical difference between v0.90 and v1.x.x that cannot be dealt with T.T

**Comment by trivialfis** [comment]
>  So I have to train model again in version 1.x.x. The issue is that I can not get same model, even if all parameters are same all data are same all seeds are same.

Sorry, due to floating point difference, you can't train the same model even without upgrading to 1.x.  I will try to explain why:
When floating point number is used, `x + y + z` is not necessarily equal to `x + (y + z)` [wiki](https://en.wikipedia.org/wiki/Associative_property#Nonassociativity_of_floating_point_calculation) .  GPU Hist builds histogram in parallel, building histogram means adding floating point values to bins.  When adding values in parallel, there's no guarantee on which value got added first.  So give original bin value `bin`, we want to add `x` and `y` to it.  In parallel, it can be `bin' = bin + x` by first thread then `bin'' = bin' + y` by second thread, or the other way around, they will generate different result.  So with or without upgrading, you are not going to get the exact same model with your current one, because parallelization can change the result every time you run.

I referred to the new parameter is that, after upgrading, your new models will be more stable.  But it won't match the model you have trained in 0.9.

**Comment by trivialfis** [comment]
Feel free to reopen or open a new issue if there's any further help I can provide.
